Additions for evolutions 4

nilsdeppe opened the pull request:
## Proposed changes

- Work around [this](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=64095) bug in GCC
- The `Tag::ElementMap` now actually holds an `ElementMap` instead of a `unique_ptr<CoordinateMapBase>`
- Add tag for computing the coordinates from a coordinate map tag and source coordinates
- Add an `DomainCreator` options tag. @wthrowe do we have a method of testing these tags that I've forgotten about?
- Fix a bug in Variables
- Changes to FluxCommunication so it interfaces with the fluxes from ScalarWave. It's likely this will need to be adapted further when we have dynamical backgrounds.

These are the last changes to FluxCommunication.hpp that I need to get evolutions to work.

**Edit:** there is one more thing I need to change in FluxCommunication.hpp, which I'm changing now. It's that dt_vars is not added by them, but rather is mutated in the DataBox.

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

nilsdeppe committed:
Make Tags::ElementMap's type be ::ElementMap

nilsdeppe committed:
Add Coordinates tag compute item

Given a  map and source coordinates in the source frame of the map
computes the coordinates in the target frame of the map.

nilsdeppe committed:
Add OptionTags::DomainCreator for reading from input file

nilsdeppe committed:
Change return type in lift_flux to be more correct

nilsdeppe committed:
Use Variables prefix tag constructor in FluxCommunication

nilsdeppe committed:
FluxCommunication no longer allocates dt_variables

nilsdeppe commented:
I've added two commits that make the remaining changes for FluxCommunication.hpp to be identical to what I have on the evolution branch.

kidder commented on src/Parallel/Main.hpp:
```diff
@@ -117,7 +117,7 @@ Main<Metavariables>::Main(CkArgMsg* msg) noexcept
         },
         [&command_line_options](std::true_type /*meta*/, auto /*mv*/,
-                                auto... /*unused*/) {
+                                ... /*unused*/) {
```
general query: does any test test the four possibilities of the enclosing make_overloader?

nilsdeppe replied:
I don't know, I'd have to look. @wthrowe might know

wthrowe replied:
Nils rewrote these tests so I'm not sure, but I think only the first case is tested.  The third case is a static_assert so that one's fine.  The fourth would be straightforward to test by creating a test with no options.  The second (the one changed here) is hard because it requires parsing an input file.  I guess it could be tested with a failure test.

kidder replied:
we can easily test this once we have an example input file parsing test

kidder commented on tests/Unit/Domain/Test_CoordinatesTag.cpp:
```diff
@@ -0,0 +1,46 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "DataStructures/DataBox.hpp"
```
add includes for size_t, Index, Frames,

kidder commented on src/DataStructures/Variables.hpp:
```diff
@@ -402,8 +402,9 @@ Variables<tmpl::list<Tags...>>::Variables(
-      number_of_grid_points_(rhs.number_of_grid_points()),
-      reference_variable_data_(std::move(rhs.reference_variable_data_)) {}
+      number_of_grid_points_(rhs.number_of_grid_points()) {
+  add_reference_variable_data(typelist<Tags...>{});
```
why didn't an existing test catch this bug?  can the test be updated?

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -72,13 +72,30 @@ struct ComputeBoundaryFlux {
+
+  template <typename Flux, typename... NumericalFluxTags, typename... Tags,
+            typename... ArgumentTags, typename... Args>
+  static auto apply_normal_dot_numerical_flux(
```
auto = void?

nilsdeppe replied:
Yep

kidder commented on tests/Unit/Domain/Test_CoordinatesTag.cpp:
```diff
@@ -0,0 +1,46 @@
+                   ElementId<Dim>(0),
+                   make_coordinate_map_base<Frame::Logical, Frame::Grid>(map)));
+  CHECK_ITERABLE_APPROX(
+      (db::get<Tags::Coordinates<map_tag, Tags::LogicalCoordinates<Dim>>>(box)),
```
will I have to fetch the coordinates this way?  I would like to get them with something like db::get<Tags::Coordinates<Frame>>(box)

nilsdeppe replied:
Yes, currently the long way is necessary. We could alternatively have things like `Tags::InertialCoordinates` etc. Another possibility might be to just template on the target frame. This would require making a canonical decision on source frame, and map. That could be a bit tricky with ElementMap and dual frames. I'd have to think about it.

kidder reviewed (changes requested):
more to come, I didn't make it all the way through

wthrowe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp:
```diff
@@ -284,10 +287,12 @@ SPECTRE_TEST_CASE(
-                                     CoordinateMaps::AffineMap>(
-          {-1., 1., 3., 7.}, {-1., 1., -2., 4.}));
+  auto map = ElementMap<2, Frame::Grid>(
+      ElementId<2>{0},
```
`self_id`.  The other one *should* also be `self_id`, but if you make that change I think the test may break.  You can check if it breaks anything if you like, but if it does just leave it and I'll fix it in my cleanup of that test.

wthrowe commented on src/Domain/Tags.hpp:
```diff
@@ -31,6 +32,17 @@ struct Logical;
+template <size_t Dim, typename TargetFrame>
+struct DomainCreator {
+  using type = std::unique_ptr<::DomainCreator<Dim, TargetFrame>>;
+  static constexpr const char* const help = {"The domain to create initially"};
```
`OptionString`

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/LiftFlux.hpp:
```diff
@@ -31,11 +31,12 @@ namespace dg {
-                          const size_t extent_perpendicular_to_boundary,
-                          DataVector magnitude_of_face_normal) noexcept {
+template <typename NormalDotFluxTags, typename NormalDotNumericalFluxTags>
+Variables<NormalDotNumericalFluxTags> lift_flux(
```
Shouldn't the return type be the time derivative of the evolved variables or something like that?

kidder replied:
`return Variabes<dt...>(std::move(num_flux))`

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -145,57 +161,51 @@ struct ComputeBoundaryFlux {
 
       // Using this instead of auto prevents incomprehensible errors
       // if the return type of compute_flux is wrong.
```
Update this comment and move it to the right place (if it is still relevant).

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -145,57 +161,51 @@ struct ComputeBoundaryFlux {
+          normal_dot_numerical_flux_computer, self_data, neighbor_data,
+          typename std::decay_t<decltype(
+              normal_dot_numerical_flux_computer)>::argument_tags{},
+          unit_face_normal);
```
Can this depend on unit_face_normal?  I thought we decided it couldn't when we were discussing fluxes in GH.

kidder replied:
just send the normal

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -110,8 +110,9 @@ struct ComputeBoundaryFlux {
-    using dt_variables_tag = typename System::dt_variables_tag;
+    using variables_tag = Tags::Variables<typename System::variables_tags>;
+    using dt_variables_tag = Tags::dt<Tags::Variables<
+        db::wrap_tags_in<Tags::dt, typename System::variables_tags>>>;
```
`using dt_variables_tag = db::add_tag_prefix<Tags::dt, variables_tag>;`

wthrowe reviewed (changes requested):
Haven't looked at the last commit.

kidder commented on src/Parallel/Main.hpp:
```diff
@@ -117,7 +117,7 @@ Main<Metavariables>::Main(CkArgMsg* msg) noexcept
         },
         [&command_line_options](std::true_type /*meta*/, auto /*mv*/,
-                                auto... /*unused*/) {
+                                ... /*unused*/) {
```
add a comment about why the change is made

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -193,18 +193,12 @@ struct ComputeBoundaryFlux {
-      // Convert tags.  This will be replaced by expanded Variables
-      // functionality.  See #319.
-      std::decay_t<decltype(dt_vars)> retagged_lifted_data = 1. * lifted_data;
+      std::decay_t<decltype(dt_vars)> lifted_data(dg::lift_flux(
```
auto after change to a previous commit

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -72,13 +72,30 @@ struct ComputeBoundaryFlux {
-                         tmpl::list<ArgumentTags...> /*meta*/) noexcept {
-    return flux(tuples::get<ArgumentTags>(self_data)...,
-                tuples::get<ArgumentTags>(neighbor_data)...);
+  template <typename... NormalDotFluxTags, typename... VariablesTags,
```
make a new action for non-conservative systems and keep the current action for conservative systems

kidder commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp:
```diff
@@ -33,10 +33,10 @@ struct Var : db::DataBoxTag {
   static constexpr const size_t volume_dim = 2;
-  using variables_tag = Tags::Variables<tmpl::list<Var>>;
-  using dt_variables_tag = Tags::dt<Tags::Variables<tmpl::list<Tags::dt<Var>>>>;
+  using variables_tags = tmpl::list<Var>;
```
write things in terms of variables_tag instead of variables_tags

kidder commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp:
```diff
@@ -122,6 +123,9 @@ SPECTRE_TEST_CASE("Unit.DiscontinuousGalerkin.Actions.FluxCommunication",
   const CoordinateMaps::AffineMap xi_map{-1., 1., 3., 7.};
   const CoordinateMaps::AffineMap eta_map{-1., 1., -2., 4.};
 
+  using dt_variables_tag = Tags::dt<Tags::Variables<
```
add_tag_prefix

kidder commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp:
```diff
@@ -279,14 +286,16 @@ SPECTRE_TEST_CASE("Unit.DiscontinuousGalerkin.Actions.FluxCommunication",
     "Unit.DiscontinuousGalerkin.Actions.FluxCommunication.NoNeighbors",
     "[Unit][NumericalAlgorithms][Actions]") {
   ActionTesting::ActionRunner<Metavariables> runner{{}};
+  using dt_variables_tag = Tags::dt<Tags::Variables<
```
add_tags_prefix

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -121,8 +121,6 @@ struct ComputeBoundaryFlux {
         db::get<Tags::HistoryBoundaryVariables<
                   Direction<volume_dim>, variables_tag>>(box);
 
-    auto dt_vars = db::item_type<dt_variables_tag>(extents.product(), 0.);
```
add the part where this is zeroed to the PR

nilsdeppe committed:
Change Systems to only have variables_tag

Removes the variables_tags. These can be computed from the
variables_tag in the cases where needed, but most things only care
about the tag anyway.

nilsdeppe commented:
I've updated this addressing the comments. The PR depends on #468. I've removed the changes to the FluxCommunication stuff and will add the new Actions in a separate PR to keep PRs smaller.

nilsdeppe committed:
Add db::data_on_slice

Similar to the Variables data_on_slice, except this will slice a
series of Tensors from the DataBox into a Variables.

nilsdeppe committed:
Move DataBox related headers to src/DataStructures/DataBox dir

kidder commented:
@wthrowe can you take a final look at this PR

wthrowe commented on src/Domain/Tags.hpp:
```diff
@@ -31,6 +33,17 @@ struct Logical;
 }  // namespace Frame
 
+namespace OptionTags {
+/// \ingroup DataBoxTagsGroup
```
Not a DataBoxTag.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/LiftFlux.hpp:
```diff
@@ -48,6 +52,8 @@ Variables<Tags> lift_flux(const Variables<Tags>& local_flux,
   auto lifted_data = std::move(numerical_flux);
```
The comment for this line is no longer correct.

nilsdeppe committed:
Add OptionsTagsGroup

nilsdeppe committed:
Move Time input options into OptionsTagGroup Doxygen group

nilsdeppe commented:
I've pushed (rebased) updates to the commits for DomainCreator tag and lift_flux return type. I also added 2 new small commits:

1. Add the OptionTagsGroup Doxygen group
2. Add the relevant tags in Time/Tags.hpp to that group. They so far were in the non-existent OptionTags group.

The rendered documentation is available [here](http://nilsdeppe.com/spectre/pr_465/)

nilsdeppe commented:
This needs to be rebased after #474 is merged.

nilsdeppe commented:
I rebased this on top of #474 so that once that is merged this can be merged too

Edit:

I did this to try deal with the long macOS build times.

kidder commented:
because of long queues for Travis mac builds, I tested this on my laptop
